framework/include/actions/ActionWarehouse.h

```diff
@@ -101,6 +101,64 @@ class ActionWarehouse :
   const std::vector<Action *> & getActionsByName(const std::string & task) const;
 
   /**
+   * Retrieve an action with its name and the desired type.
+   * @param name The action name.
+   */
+  template <class T>
+  const T & getAction(const std::string & name)
+    {
+      T* p = NULL;
+      for (std::vector<std::string>::iterator it = _ordered_names.begin(); it != _ordered_names.end(); ++it)
+      {
+        std::string task = *it;
+        for (ActionIterator act_iter = actionBlocksWithActionBegin(task);
+             act_iter != actionBlocksWithActionEnd(task);
+             ++act_iter)
+        {
+          if ((*act_iter)->name() == name)
+          {
+            p = dynamic_cast<T*>(*act_iter);
+            if (!p)
+              mooseError("The action with name being "+name+" is not in the desired type");
+            else
+              break;
+          }
+        }
+      }
+      if (!p)
+        mooseError("Action with name being "+name+" does not exist");
+      return *p;
+    }
+
+  /**
+   * Retrieve all actions in a specific type.
+   */
+  template <class T>
+  std::vector<const T *> getActions()
```

Comment:
I will keep the name but return std::set, then developers will know this is returning unique actions. I do not quite like 'unique' because it is confusing and people will wondering what and why is unique. I did not mention test and was afraid you will raise the question and you did... How do we test this any idea? I have tested this with Yak, but I guess you want something in MooseTest.